docs: describe how to manage embedded plugin registry for VS Code editor (#2503)
codenvy/che-docs: modules/administration-guide/pages/extensions-for-microsoft-visual-studio-code-open-source.adoc

[id="extensions-for-microsoft-visual-studio-code-open-source"]
= Extensions for Microsoft Visual Studio Code - Open Source

> To manage extensions for link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE], {prod-short} uses one of these link:https://open-vsx.org/about[Open VSX] registries:

Reviewer: Che doesn't manage extensions (i.e. it doesn't know what a VS Code extension is). VS Code itself manages the extensions. Che configures the Open VSX registry that VS Code will connect to.   

```suggestion
To manage extensions, the link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE] cloud editors use one of these link:https://open-vsx.org/about[Open VSX] registries:
```

* The public, primary link:https://open-vsx.org[open-vsx.org] registry.
* The embedded, locally cached instance of the Open VSX registry that runs in the `plugin-registry` pod of {prod-short} to support air-gapped, offline, and proxy-restricted environments. The embedded Open VSX registry contains only a subset of the extensions published on link:https://open-vsx.org[open-vsx.org].

> * Your own custom build of the embedded Open VSX registry for use in your organization's workspaces.

Reviewer: Second and third bullets are both "embedded Open VSX registry that runs in the `plugin-registry`": we should merge them in one bullet point and mention that the subset of available extensions can be customized.

I would add a last Open VSX registry option: a standalone instance. This is distinct from the second bullet because it's not running within the `plugin-registry` pod and is managed separately from the Che instance. And it's also distinct from the first bullet point because it's a customer managed instance, distinct from the one managed by the Eclipse foundation at open-vsx.org.

```suggestion
* The embedded instance of the Open VSX registry that runs in the `plugin-registry` pod of {prod-short} to support air-gapped, offline, and proxy-restricted environments. The embedded Open VSX registry contains only a subset of the extensions published on link:https://open-vsx.org[open-vsx.org]. This subset can be customized.
* A standalone Open VSX registry instance, deployed on a network accessible from {prod-short} workspace pods.
```

[id="extensions-for-microsoft-visual-studio-code-open-source"]
= Extensions for Microsoft Visual Studio Code - Open Source

> To manage extensions for link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE], {prod-short} uses one of the link:https://open-vsx.org/about[Open VSX] registry instances:

Reviewer 2: ```suggestion
To manage extensions for link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE], {prod-short} uses one of these link:https://open-vsx.org/about[Open VSX] registries:
```

Reviewer 3: There can be more than three, given the unlimited third point, no?

Reviewer 2: would you prefer "these three CATEGORIES of registry instances" ?

or just "these registries" ?

Reviewer 3: @nickboldt, thank you, Nick. These registries work fine. I'm committing your suggestion.

> * Public link:https://open-vsx.org[open-vsx.org].

Reviewer 2: ```suggestion
* The public, primary link:https://open-vsx.org[open-vsx.org] registry.
```

* The embedded, locally cached instance of the Open VSX registry that runs in the `plugin-registry` pod of {prod-short} to support air-gapped, offline, and proxy-restricted environments, containing only a subset of the extensions published on link:https://open-vsx.org[open-vsx.org].

> * Your own custom build of the embedded Open VSX registry for use in workspaces of your organization's {prod-short}.

Reviewer 2: ```suggestion
* Your own custom build of the embedded Open VSX registry, for use in your organization's workspaces.
```

> The Open VSX registry instance default differs depending on how proxies or DNS are configued: If {prod-short} resolves `pass:c,a,q[https://open-vsx.org]`, then the default is link:https://open-vsx.org[open-vsx.org]. If not, then the default is the embedded, locally cached Open VSX registry instance within the `plugin-registry` pod of {prod-short}.

Reviewer 2: ```suggestion
If it can be resolved from within your cluster, the default Open VSX registry at `pass:c,a,q[https://open-vsx.org]` will be used; if not, default is the embedded, locally cached Open VSX registry within the {prod-short} `plugin-registry` pod.
```

Reviewer 3: @nickboldt, thank you, Nick. I'm committing your suggestion.

[id="extensions-for-microsoft-visual-studio-code-open-source"]
= Extensions for Microsoft Visual Studio Code - Open Source

To manage Visual Studio Code extensions {prod-short} uses link:https://open-vsx.org/[Open VSX Registry], it is an open source VS Code extensions registry.

> Depends on which cluster is used, with a restricted/disconnected network (air gap) or a connected network, the default value of Open VSX registry is different. In a cluster with a restricted network it is a local Open VSX registry instance within the plugin-registry pod, in another case - it is a public link:https://open-vsx.org/[Open VSX Registry].

Reviewer 2: ```suggestion
The default Open VSX registry differs depending on how proxies or DNS is configued. 

If open-vsx.org can be resolved, the default will be to the public link:https://open-vsx.org/[Open VSX Registry]; if not, then the default will be a locally cached Open VSX registry instance within the plugin-registry pod to support airgapped, offline, and proxy-restricted deployments. 
```

> It is also possible to control what instance of Open VSX should be used in the workspaces. This configuration is part of the `CheCluster` Custom Resource:

Reviewer 2: ```suggestion
It is also possible to control what instance of Open VSX should be used in the workspaces, should you want to include your own custom build. This configuration is part of the `CheCluster` Custom Resource:
```

[source,yaml,subs="+quotes"]
----
spec:
  components:
    pluginRegistry:
      openVSXURL: "__<Open_VSX_registry_URL>__" # <1> 
----

> <1> Sets the URL of Open VSX registry. The default value is `https://open-vsx.org`. In case if the value is empty, {prod-short} will use a local Open VSX registry.

Reviewer 2: ```suggestion
<1> Sets the URL of Open VSX registry. The default value is `https://open-vsx.org`. An empty string will cause  {prod-short} to use the local Open VSX registry in the plug-in registry pod.
```

> The local/embedded registry is running in the plugin-registry pod and provides a subset of VS Code extensions that are published on link:https://open-vsx.org/[Open VSX Registry].

Reviewer 2: ```suggestion
The local Open VSX registry runs in the plug-in registry pod and provides a subset of the VS Code extensions published on link:https://open-vsx.org/[Open VSX Registry].
```

> .Next topics describe how to manage a local Open VSX registry:

Reviewer 2: ```suggestion
.Below, please find information about managing a local Open VSX registry:
```

[id="extensions-for-microsoft-visual-studio-code-open-source"]
= Extensions for Microsoft Visual Studio Code - Open Source

> To manage Visual Studio Code extensions {prod-short} uses link:https://open-vsx.org/[Open VSX Registry], it is an open source VS Code extensions registry.

Reviewer 2: ```suggestion
To manage extensions for link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE], {prod-short} uses the link:https://open-vsx.org/[Open VSX Registry]. 

This is an open source extensions registry, restricted in its use like the link:https://marketplace.visualstudio.com/[Visual Studio Code Marketplace] which can only be used with link:https://aka.ms/vsmarketplace-ToU[In-Scope Products and Services from Microsoft].
```

The default Open VSX registry differs depending on how proxies or DNS is configued. 

> If open-vsx.org can be resolved, the default will be to the public link:https://open-vsx.org/[Open VSX Registry]; if not, then the default will be a locally cached Open VSX registry instance within the plugin-registry pod to support airgapped, offline, and proxy-restricted deployments. 

Reviewer 2: ```suggestion
If open-vsx.org can be resolved, the default will be to the public link:https://open-vsx.org/[Open VSX Registry]; if not, then the default will be a locally cached Open VSX registry instance within the plugin-registry pod to support air-gapped, offline, and proxy-restricted deployments. 
```

It is also possible to control what instance of Open VSX should be used in the workspaces, should you want to include your own custom build. This configuration is part of the `CheCluster` Custom Resource:

[source,yaml,subs="+quotes"]
----
spec:
  components:
    pluginRegistry:
      openVSXURL: "__<Open_VSX_registry_URL>__" # <1> 
----

<1> Sets the URL of Open VSX registry. The default value is `https://open-vsx.org`. An empty string will cause  {prod-short} to use the local Open VSX registry in the plug-in registry pod.

> The local Open VSX registry runs in the plug-in registry pod and provides a subset of the VS Code extensions published on link:https://open-vsx.org/[Open VSX Registry].

Reviewer 2: ```suggestion
The local Open VSX registry runs in the plug-in registry pod and provides a subset of the Visual Studio Code extensions published on link:https://open-vsx.org/[Open VSX Registry].
```

> .Below, please find information about managing a local Open VSX registry:

Reviewer 2: rather than a call to action like "please find" should this be a section title called "managing a local Open VSX registry" ?

[id="extensions-for-microsoft-visual-studio-code-open-source"]
= Extensions for Microsoft Visual Studio Code - Open Source

To manage extensions for link:https://github.com/microsoft/vscode[Microsoft Visual Studio Code - Open Source IDE], {prod-short} uses the link:https://open-vsx.org/[Open VSX Registry]. 

This is an open source extensions registry, restricted in its use like the link:https://marketplace.visualstudio.com/[Visual Studio Code Marketplace] which can only be used with link:https://aka.ms/vsmarketplace-ToU[In-Scope Products and Services from Microsoft].

> The default Open VSX registry differs depending on how proxies or DNS is configued. 

Reviewer: ```suggestion
The default Open VSX registry differs depending on how proxies or DNS are configured. 
```
